pkg/manifest/jsonmanifest/jsonmanifest_test.go

```diff
@@ -0,0 +1,243 @@
+	}
+
+	m := jsonmanifest.NewManifest()
+	if m.Length() != 0 {
+		t.Fatalf("expected length to be %d, but is %d instead", 0, m.Length())
+	}
+
+	// add and check all entries
+	for i, e := range tc.entries {
+		entry := jsonmanifest.NewEntry(
+			e.reference,
+			e.name,
+			e.headers,
+		)
+		path := filepath.Join(e.path, e.name)
+		m.Add(path, entry)
+
+		// verify length change
+		if m.Length() != i+1 {
+			t.Fatalf("expected length to be %d, but is %d instead", i+1, m.Length())
+		}
+		// check retrieved entry
+		verifyEntry(t, m, entry, path)
+	}
+	// save current manifest length
+	manifestLen := m.Length()
+
+	// create new entry to replace existing one
+	lastEntry := tc.entries[len(tc.entries)-1]
+	entry := jsonmanifest.NewEntry(
+		test.RandomAddress(),
+		lastEntry.name,
+		lastEntry.headers,
+	)
+
+	// replace manifest entry by adding to the same path
+	path := filepath.Join(lastEntry.path, lastEntry.name)
+	m.Add(path, entry)
+
+	// length should not have changed
+	if m.Length() != manifestLen {
+		t.Fatalf("expected length to be %d, but is %d instead", manifestLen, m.Length())
+	}
+	// check retrieved entry
+	verifyEntry(t, m, entry, path)
+
+	// try removing inexistent entry
+	m.Remove("invalid/path.ext")
+
+	// length should not have changed
+	if m.Length() != manifestLen {
+		t.Fatalf("expected length to be %d, but is %d instead", manifestLen, m.Length())
+	}
+
+	// remove each entry
+	for i, e := range tc.entries {
+		path := filepath.Join(e.path, e.name)
+		m.Remove(path)
+
+		// verify length change
+		if m.Length() != manifestLen-i-1 {
+			t.Fatalf("expected length to be %d, but is %d instead", manifestLen-i-1, m.Length())
+		}
+	}
+}
+
+// verifyEntry checks that an entry is equal to the one retrieved from the given manifest and path.
+func verifyEntry(t *testing.T, m manifest.Interface, entry manifest.Entry, path string) {
+	re, err := m.Entry(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(entry, re) {
+		t.Fatalf("original and retrieved entry are not equal: %v, %v", entry, re)
+	}
+}
+
+// TestEntryModification verifies that manifest entries are not modifiable from outside of the manifest.
+func TestEntryModification(t *testing.T) {
+	m := jsonmanifest.NewManifest()
+
+	// add single entry
+	e := jsonmanifest.NewEntry(
+		test.RandomAddress(),
+		"single_entry.png",
+		http.Header{"Content-Type": {"image/png"}},
+	)
+	m.Add("", e)
+
+	// retrieve entry
+	re, err := m.Entry("")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// modify entry
+	re.Headers().Add("Content-Type", "text/plain; charset=utf-8")
+
+	// re-retrieve entry and compare
+	rre, err := m.Entry("")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if reflect.DeepEqual(rre, re) {
+		t.Fatalf("manifest entry %v was unexpectedly modified externally", rre)
+	}
+}
+
+// TestMarshal verifies that created manifests are successfully marshalled and unmarshalled.
+func TestMarshal(t *testing.T) {
+	for _, tc := range testCases {
```

Comment:
you don't need multiple test cases for marshaling. one test will do

Reply:
i'm not so convinced. a few commits ago, a tc with `nil` entries for the manifest was failing the marshalling, while _all_ the rest were succeeding.

we have the following cases:
- nil entries
- 1 plain entry
- 2 plain entries
- 4 nested entries 

i tried to think of representative test cases here. maybe i went a bit overboard, but i don't think it's doing more bad than good at this point.

also the last of these is actually used for another test, which verifies the `Add`, `Remove`, `Entry` and `Length` functions, so i think it should have at least _some_ complexity.

Reply:
the (unmarshal) just calls json marshal. not sure why we need 4 tests to see that this builtin go functionality works. it is true that it comes at a low price in this case, but you could probably have 1 test case that covers everything. also, for `nil` entries it is not clear to me how and why such manifest would even be created. this should potentially not do anything or return an error

Reply:
> the (un)marshal just calls json marshal. not sure why we need 4 tests to see that this builtin go functionality works. it is true that it comes at a low price in this case, but you could probably have 1 test case that covers everything. 

ultimately, you're right. the thing is that i coded these 4 test cases when the (un)marshaling functions were not just built-in go functionalities, but had some logic instead (unexported structs with json annotations and yada yada). 

after this, you, @janos and @pradovic helped me out in simplifying the structs (and their fields) so that we ended up not really needing anything else but the `json.Marshal` calls.

since the nil entries test-case is an interesting edge case (i will go into this shortly) and the last test case is used for testing the `Add`, `Remove`, `Entry` and `Length` funcs, we really only have 2 more test cases than what we really _need_; i.e. `one-entry` and `two-entries`.

> also, for `nil` entries it is not clear to me how and why such manifest would even be created. this should potentially not do anything or return an error

a newly created manifest starts out empty (meaning, no entries), so this can for sure exist in the code. whether it _should_ though, is a different story.

as far as i know, we don't (currently) have a place in the code which outputs such result, but i—maybe prematurely—anticipated this for issue #481. it is my understanding that an empty directory should have no entries; but right now, if you upload a tar with no files in it, this will result in an error. so yes, there is some inconsistency here that should be dealt with.

i hope to settle this when we tackle #481 or continue to iterate on manifests. i will keep it in mind—if it turns out that there is no good reason to keep an empty manifest around, then for sure let's have this as an error case.